File: src/main/java/ux/ToggleHighlightControl.java
package ux;

import actions.KeyCallback;
import actions.KeyObservable;

public class ToggleHighlightControl extends Control {
    protected ToggleHighlightControl(Builder builder) {
        super(builder);
    }

    @Override
    public void onMouseReleased() {
        super.onMouseReleased();
        isHighlightFromKeypress = !isHighlightFromKeypress;
    }

    @Override
    public void notifyKeyPress(KeyObservable o) {
        // Specific behavior for ToggleHighlightControl
        isHighlightFromKeypress = !isHighlightFromKeypress;
    }

    public static class Builder extends Control.Builder {

        public Builder(DrawingInfoSupplier drawingInformer, KeyCallback callback, String iconName, int size) {
            super(drawingInformer, callback, iconName, size);
        }

        @Override
        public ToggleHighlightControl build() {
            return new ToggleHighlightControl(this);
        }

        @Override
        public Builder self() {
            return this;
        }

    }
}

File: src/main/kotlin/ux/panel/ToggleHighlightControl.kt
package ux.panel

import actions.KeyCallback
import actions.KeyObservable
import ux.informer.DrawingInfoSupplier

class ToggleHighlightControl private constructor(builder: Builder?) : Control(
    builder!!
) {
    override fun onMouseReleased() {
        super.onMouseReleased()
        isHighlightFromKeypress = !isHighlightFromKeypress
    }

    override fun notifyKeyPress(observer: KeyObservable) {
        // Specific behavior for ToggleHighlightControl
        isHighlightFromKeypress = !isHighlightFromKeypress
    }

    class Builder(drawingInformer: DrawingInfoSupplier?, callback: KeyCallback?, iconName: String?, size: Int) :
        Control.Builder(drawingInformer, callback!!, iconName!!, size) {
        override fun build(): ToggleHighlightControl {
            return ToggleHighlightControl(this)
        }

        override fun self(): Builder {
            return this
        }
    }
}